### ISSUE TITLE
EDN formatted output is not round-trippable

### ISSUE THREAD

**Issue by yuhan0** [clj-kondo/clj-kondo#2899]
- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**version**

clj-kondo v2026.05.25

**macro usage**

NA

**platform**

native (macOS)

**editor**

NA

**problem**

Running clj-kondo with config `{:output {:format :edn}}` emits findings via `prn` on the raw finding maps with no coercion/validation. 

Several fields in `:findings` contain symbols _taken verbatim from user source_ - this can result in illegal EDN output that downstream consumers cannot read back.

**repro**

```bash
$ echo '::foo/1a' | clj-kondo --lint - --config '{:output {:format :edn}}' | bb -e '(clojure.edn/read *in*)'

----- Error --------------------------------------------------------------------
Type:     clojure.lang.EdnReader$ReaderException
Message:  java.lang.NumberFormatException: Invalid number: 1a
Location: NO_SOURCE_PATH:1:1

----- Context ------------------------------------------------------------------
1: (clojure.edn/read *in*)
   ^--- java.lang.NumberFormatException: Invalid number: 1a

----- Stack trace --------------------------------------------------------------
clojure.edn/read - clojure/edn.clj:14:1
user             - NO_SOURCE_PATH:1:1
```

**config**

NA

**expected behavior**

EDN output should be round-trippable: `clj-kondo --config {:output :edn}` → `clojure.edn/read` guaranteed to never throw.

### Investigation

I've found two distinct ways broken EDN can occur (may not be exhaustive):

**1. Problematic symbol coercion**

Certain ident segments starting with digits `:1a/b` or `:099` are (de facto) valid as keywords, but will be read back as invalid numbers when broken apart and treated as symbols. 
Similarly `#` is a valid keyword-starter but not symbol (`:#id` (valid) → `{:name #id}` will be read as a broken tagged literal) 

Also consider various silently non-throwing but semantically wrong situations like `:nil/true` → `{:ns nil :name true}` / `:#_/123` → `{:ns #_ :name 123}`

**2. Invalid ident passthrough**

The rewrite-clj parser is more lenient than the actual Clojure reader, representing illegal tokens like `foo/` as flagged AST nodes instead of exiting early. 

Downstream linters may naively interpolate these nodes into their findings, e.g.

```clojure
(ns foo
  (:require [bar/ :refer [qux:]]))
```

In addition to invalid-symbol errors, this also triggers the `:unused-referred-var` linter which produces the (malformed EDN) findings map

```clojure
 {:type :unused-referred-var, 
  :level :warning, 
  :message "#'bar//qux: is referred but never used"  
  :referred-ns bar/
  :refer qux:
  ...}
```
 
---

It's not immediately obvious where the fix or proper validation/normalization boundary should lie - 'hardening' the EDN writer by stringifying the relevant value fields  seems like a sensible change, but that would be a schema break.

Note: `:edn` is the only affected format - `:json` and `:sarif` stringify all symbols to strings, and `:text` is unstructured.

**Comment by borkdude** [comment]
Maybe we should just use strings for symbols but it would be a breaking change I guess.

**Comment by yuhan0** [comment]
Yeah.. a minor one I suppose, but up for you to make the call? 

For a bit more context, I ran into this while attempting to programmatically diff clj-kondo's output with the `RELEASE` version on a collection of messy scratchpad-type fiddle files (2600+ findings total) The actual culprit was a spurious `gauge/1x1` that had gotten uncommented.

Pointed claude at the codebase too, here's a list of linters according to it which may trigger the issue:

| linter | field(s) | repro |
|---|---|---|
| `:unresolved-namespace` | `:ns`, `:name` | `::foo/1a`, `x/`, `a//b` |
| `:unused-namespace` | `:ns` | `(ns x (:require [foo/ :refer [a]]))` |
| `:unused-referred-var` | `:ns`, `:referred-ns`, `:refer` | `(ns x (:require [foo/ :refer [a]]))` |
| `:missing-protocol-method` | `:protocol-ns`, `:protocol-name` | (uncertain) |
| `:refer-all`, `:use` | `:refers` | (uncertain) |

(where it couldn't construct a trigger for the latter two, treat as non-exhaustive) 
It also gave the verdict that other species of malformed-edn besides symbols (like regexes, `#object[...]`, `{a 1, a 2}`) were currently impossible by construction

**Comment by NoahTheDuke** [comment]
> EDN output should be round-trippable: clj-kondo --config {:output :edn} → clojure.edn/read guaranteed to never throw.

I don't think this is necessarily true. clj-kondo works on broken/illegal code,  why would it be expected to output legal code?

**Comment by borkdude** [comment]
One creative solution could be to have `#symbol` data reader or so, but this would also require people to use that. I don't think there is a non-breaking path. If you're using claude, I'm sure you can teach it to fix broken EDN ;).

**Comment by yuhan0** [comment]
@NoahTheDuke I think that conflates two different notions of 'legal' - the whole point of a linter should be to transform broken code into intelligible findings - if I'm asking clj-kondo for structured `:json` / `:edn` output surely the minimum contract should be for it to be machine-readable? (claude usage not relevant to the discussion - I just brought it up to surface more context)

**Comment by NoahTheDuke** [comment]
hmm yeah, i guess that's true. what does it output when analyzing `{:a :b]`?

**Comment by borkdude** [comment]
I guess we could make an opt-in setting to sanitize the outputted EDN and use either strings or a reader tag in front of a string to make it into the correct symbol/keyword (with spaces, etc).

**Comment by NoahTheDuke** [comment]
i like the reader tag idea. how many things are readable but not round trip-able? keywords, symbols, numbers?
